File: Algorithms/p015_3Sum/p015_3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        n = len(nums)
        res = []
        for i in range(n-2):
            if i>0 and nums[i]==nums[i-1]:
                continue
            l = i+1
            r = n-1
            while l<r:
                if nums[i]+nums[l]+nums[r]<0:
                    l += 1
                elif nums[i]+nums[l]+nums[r]>0:
                    r -= 1
                else:
                    res.append([nums[i],nums[l],nums[r]])
                    l += 1
                    r -= 1
                    while l<r and nums[l]==nums[l-1]:
                        l += 1
                    while l<r and nums[r]==nums[r+1]:
                        r -= 1
        return res

File: Algorithms/p015_3Sum/p015_3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        int n = nums.size();
        vector<vector<int>> res;
        for(int i=0;i<n-2;++i){
            if(i>0&&nums[i]==nums[i-1]) continue;
            int l = i+1;
            int r = n-1;
            while(l<r){
                if(nums[i]+nums[l]+nums[r]<0){
                    ++l;
                }else if(nums[i]+nums[l]+nums[r]>0){
                    --r;
                }else{
                    res.push_back({nums[i],nums[l],nums[r]});
                    ++l;
                    --r;
                    while(l<r&&nums[l]==nums[l-1]) ++l;
                    while(l<r&&nums[r]==nums[r+1]) --r;
                }
            }
        }
        return res;
    }
};
